### PR TITLE
Implemented editBusinessRule related methods, save businessRule to db and minor fixes

### PR DESCRIPTION
**Pull request by AnuGayan**

### CODE REVIEW

File: components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/util/TemplateManagerConstants.java [hunk 1]
```diff
@@ -18,19 +18,23 @@
...
+import org.wso2.carbon.utils.Utils;
+
 /**
  * Consists of constants related to root.Template Manager
  */
 public class TemplateManagerConstants {
     // Directory locations
-    public static final String TEMPLATES_DIRECTORY = "/home/senthuran/Desktop/rough-templates/"; // todo: not finalized
+//    public static final String CARBON_HOME = Utils.getCarbonHome().toString();
+    //    public static final String TEMPLATES_DIRECTORY = CARBON_HOME+"/resources/businessRules/templates/"; // todo: not
```
Comment (ramindu90) [inline components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/util/TemplateManagerConstants.java]: Is this necessary?

Comment (AnuGayan) [reply components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/util/TemplateManagerConstants.java]: This is the required one. I just commented it to test the method locally without packing with SP

Comment (ramindu90) [reply components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/util/TemplateManagerConstants.java]: @AnuGayan, shall we remove this?

Comment (AnuGayan) [reply components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/util/TemplateManagerConstants.java]: This was fixed by 47e9832228beaa5747e074303ac9d729cc69abdf and  966ec96dbd7f65065ffae47d756cafd9ef841718

File: components/org.wso2.carbon.business.rules.core/src/test/java/org/wso2/carbon/business/rules/core/TemplateManagerTest.java [hunk 2]
```diff
@@ -1,14 +1,26 @@
 package org.wso2.carbon.business.rules.core;
```
Comment (mohanvive) [inline components/org.wso2.carbon.business.rules.core/src/test/java/org/wso2/carbon/business/rules/core/TemplateManagerTest.java]: What does this class do? 
No any license header as well

File: components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/services/TemplateManagerService.java [hunk 3]
```diff
@@ -91,7 +96,8 @@ public void createBusinessRuleFromTemplate(BusinessRuleFromTemplate businessRule
                 saveBusinessRuleDefinition(businessRuleUUID, businessRuleFromTemplate, isDeployed); // todo: implement method
             } catch (TemplateManagerException e) {
                 // Saving definition is unsuccessful
-                log.error("Error in saving the Business Rule definition", e); // Exception is thrown from the saveBusinessRuleDefinition method itself
+                log.error("Error in saving the Business Rule definition", e);
+                // Exception is thrown from the saveBusinessRuleDefinition method itself
             }
         } catch (Exception e) {
             log.error(e.getMessage(), e);
```
Comment (dnwick) [inline components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/services/TemplateManagerService.java]: Give context and why catching exception ?

File: components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/services/TemplateManagerService.java [hunk 4]
```diff
@@ -118,39 +124,65 @@ public void createBusinessRuleFromScratch(BusinessRuleFromScratch businessRuleFr
...
             // Overwriting definition / Update Deploy unsuccessful
             log.error(e.getMessage(), e);
         }
     }
+    public void editBusinessRuleFromScratch(String uuid, BusinessRuleFromScratch businessRuleFromScratch) {
+        Map<String,Artifact> derivedArtifacts = null;
+        boolean isDeployed =false;
+        try {
+            derivedArtifacts = deriveArtifacts(businessRuleFromScratch);
+
```
Comment (dnwick) [inline components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/services/TemplateManagerService.java]: format properly

File: components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/services/TemplateManagerService.java [hunk 5]
```diff
@@ -196,15 +228,21 @@ public void deleteBusinessRule(String uuid) throws TemplateManagerException { //
                     removeBusinessRuleDefinition(uuid);
                 } catch (TemplateManagerException e) {
                     log.error("Failed to delete Business Rule definition of : " + uuid, e);
+                } catch (SQLException e) {
+                    e.printStackTrace();
```
Comment (dnwick) [inline components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/services/TemplateManagerService.java]: Remove and add log.error with context

File: components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/services/TemplateManagerService.java [hunk 6]
```diff
@@ -196,15 +228,21 @@ public void deleteBusinessRule(String uuid) throws TemplateManagerException { //
...
+                } catch (BusinessRulesDatasourceException e) {
+                    e.printStackTrace();
```
Comment (dnwick) [inline components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/services/TemplateManagerService.java]: Remove and add log.error with context

File: components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/services/TemplateManagerService.java [hunk 7]
```diff
@@ -220,17 +258,29 @@ public boolean deployBusinessRule(BusinessRuleFromTemplate businessRuleFromTempl
...
         return true;
 
     }
+    public boolean updateBusinessRule(BusinessRuleFromScratch businessRuleFromScratch) throws TemplateManagerException {
+        Map<String,Artifact> derivedTemplates = deriveArtifacts(businessRuleFromScratch);
+        Artifact deployableSiddhiApp = buildSiddhiAppFromScratch(derivedTemplates,businessRuleFromScratch);
+        boolean isDeployed;
+        try {
+           isDeployed = updateDeployedSiddhiApp(businessRuleFromScratch.getUuid(),deployableSiddhiApp);
+        }catch (TemplateManagerException e){
```
Comment (dnwick) [inline components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/services/TemplateManagerService.java]: format

File: components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/services/TemplateManagerService.java [hunk 8]
```diff
@@ -736,18 +809,37 @@ public RuleTemplate getRuleTemplate(BusinessRuleFromTemplate businessRuleFromTem
...
-     * @throws TemplateManagerException
+     * @throws TemplateManagerException,UnsupportedEncodingException,BusinessRulesDatasourceException,SQLException
      */
     public void saveBusinessRuleDefinition(String uuid, BusinessRuleFromTemplate businessRuleFromTemplate, boolean
             isDeployed) throws
-            TemplateManagerException {
+            TemplateManagerException, UnsupportedEncodingException, BusinessRulesDatasourceException, SQLException {
+        QueryExecutor queryExecutor = new QueryExecutor();
+        int deploymentState = 0;
+        byte[] businessRule=businessRuleFromTemplate.toString().getBytes("UTF-8");
```
Comment (dnwick) [inline components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/services/TemplateManagerService.java]: Format the code properly !!

File: components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/services/TemplateManagerService.java [hunk 9]
```diff
@@ -792,8 +887,28 @@ public void deploySiddhiApp(String siddhiAppName, Artifact siddhiApp) throws Tem
...
-    public void overwriteBusinessRuleDefinition(String uuid, BusinessRuleFromTemplate businessRuleFromTemplate) throws TemplateManagerException {
-        //todo: implement
+    public void overwriteBusinessRuleDefinition(String uuid, BusinessRuleFromTemplate businessRuleFromTemplate, boolean
+            isDeployed) throws TemplateManagerException, UnsupportedEncodingException, BusinessRulesDatasourceException,
+            SQLException {
+        QueryExecutor queryExecutor = new QueryExecutor();
+        int deploymentState = 0;
+        byte[] businessRule = businessRuleFromTemplate.toString().getBytes("UTF-8");
+        if (isDeployed){
+            deploymentState=1;
```
Comment (dnwick) [inline components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/services/TemplateManagerService.java]: Format please

File: components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/services/TemplateManagerService.java [hunk 10]
```diff
@@ -868,7 +991,8 @@ public void undeployTemplate(String templateType, String uuid) throws TemplateMa
      * @throws TemplateManagerException
      */
     public void undeploySiddhiApp(String uuid) throws TemplateManagerException {
-        // todo: implement
+        SiddhiAppApiHelper siddhiAppApiHelper = new SiddhiAppApiHelper();
+        siddhiAppApiHelper.delete("localhost",uuid);
```
Comment (dnwick) [inline components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/services/TemplateManagerService.java]: Why we need a host ? and why its hardcoded ? if needed use constants

File: components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/util/TemplateManagerHelper.java [hunk 11]
```diff
@@ -230,20 +234,23 @@ public static void validateRuleTemplate(RuleTemplate ruleTemplate) throws Templa
...
                             ruleTemplate.getUuid());
                 }
             }
             if (!(ruleTemplate.getType().equals(TemplateManagerConstants.RULE_TEMPLATE_TYPE_TEMPLATE) ||
                     ruleTemplate.getType().equals(TemplateManagerConstants.RULE_TEMPLATE_TYPE_INPUT) ||
                     ruleTemplate.getType().equals(TemplateManagerConstants.RULE_TEMPLATE_TYPE_OUTPUT))) {
-                throw new TemplateManagerException("Invalid rule template - invalid rule template type for rule template " +
+                throw new TemplateManagerException("Invalid rule template - " +
+                        "invalid rule template type for rule template " +
                         "" + ruleTemplate.getUuid());
```
Comment (dnwick) [inline components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/util/TemplateManagerHelper.java]: unnecessary new line format properly

File: components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/datasource/QueryExecutor.java [hunk 12]
```diff
@@ -39,9 +39,12 @@ public QueryExecutor() {
         queryManager = DataHolder.getInstance().getQueryManager();
     }
 
-    public boolean executeInsertQuery(String uuid, Blob businessRule, int deploymentStatus) throws BusinessRulesDatasourceException, SQLException {
+    public boolean executeInsertQuery(String uuid, byte[] businessRule, int deploymentStatus) throws
+            BusinessRulesDatasourceException, SQLException {
         Connection conn = dataSource.getConnection();
-        PreparedStatement statement = getInsertQuery(conn,uuid, businessRule, deploymentStatus);
+        Blob businessRuleBlob = conn.createBlob();
+        businessRuleBlob.setBytes(1,businessRule);
```
Comment (minudika) [inline components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/datasource/QueryExecutor.java]: There should be a space after the comma

File: components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/datasource/QueryExecutor.java [hunk 13]
```diff
@@ -51,13 +54,17 @@ public boolean executeDeleteQuery(String uuid) throws SQLException, BusinessRule
         return statement.execute();
     }
 
-    public boolean executeUpdateBusinessRuleQuery(String uuid, Blob newBusinessRule, int deploymentStatus) throws SQLException, BusinessRulesDatasourceException {
+    public boolean executeUpdateBusinessRuleQuery(String uuid, byte[] newBusinessRule, int deploymentStatus) throws
+            SQLException, BusinessRulesDatasourceException {
         Connection conn = dataSource.getConnection();
-        PreparedStatement statement = getUpdateBusinessRuleQuery(conn, uuid, newBusinessRule, deploymentStatus);
+        Blob newBusinessRuleBlob = conn.createBlob();
+        newBusinessRuleBlob.setBytes(1,newBusinessRule);
```
Comment (minudika) [inline components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/datasource/QueryExecutor.java]: Check formatting. Also do the same for other places as well.

File: components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/services/TemplateManagerService.java [hunk 14]
```diff
@@ -118,39 +124,65 @@ public void createBusinessRuleFromScratch(BusinessRuleFromScratch businessRuleFr
...
             log.error(e.getMessage(), e);
         }
 
         try {
-            overwriteBusinessRuleDefinition(uuid, businessRuleFromTemplate);
             // Load all available Business Rules again
             this.availableBusinessRules = loadBusinessRules();
             // Update Deploy templates, only if overwriting Business Rule Definition is successful
             // todo: (Q) is this ok?
             for (String templateUUID : derivedTemplates.keySet()) {
```
Comment (ksdperera) [inline components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/services/TemplateManagerService.java]: Please validate whether we need to continue this loop for rest of the key, if there is exception in one key due to any of the issues( 'TemplateManagerException | UnsupportedEncodingException | SQLException | BusinessRulesDatasourceException )

File: components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/services/TemplateManagerService.java (hunk 5 above)
Comment (ksdperera) [reply components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/services/TemplateManagerService.java]: @AnuGayan Please prioritize the fixes like this and add relevant commit ID when you reply.

File: components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/services/TemplateManagerService.java (hunk 6 above)
Comment (ksdperera) [reply components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/services/TemplateManagerService.java]: @AnuGayan Please prioritize the fixes like this and add relevant commit ID when you reply.

File: components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/services/TemplateManagerService.java [hunk 15]
```diff
@@ -504,9 +563,13 @@ public RuleTemplate getRuleTemplate(String templateGroupUUID, String ruleTemplat
     }
 
 
-    private Artifact buildSiddhiAppFromScratch(Map<String, Artifact> derivedTemplates, BusinessRuleFromScratch businessRuleFromScratch)
+    private Artifact buildSiddhiAppFromScratch(Map<String, Artifact> derivedTemplates,
+                                               BusinessRuleFromScratch businessRuleFromScratch)
             throws
             TemplateManagerException {
+        ClassLoader classLoader = TemplateManagerService.class.getClassLoader();
+        String SIDDHI_APP_TEMPLATE = classLoader.getResource("siddhi-app-template.json").getFile();
```
Comment (ksdperera) [inline components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/services/TemplateManagerService.java]: Please use constant for file name "siddhi-app-template.json"

File: components/org.wso2.carbon.business.rules.core/src/test/java/org/wso2/carbon/business/rules/core/TemplateManagerTest.java (hunk 2 above)
Comment (AnuGayan) [reply components/org.wso2.carbon.business.rules.core/src/test/java/org/wso2/carbon/business/rules/core/TemplateManagerTest.java]: This was used for testing purposes while doing the implementation, removed through 1dd71e2f12ee08489377cc42b1df7f942286e0ae

File: components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/services/TemplateManagerService.java (hunk 3 above)
Comment (AnuGayan) [reply components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/services/TemplateManagerService.java]: FIxed by 9cf9a377bbb4bb9a29b5e04ffc9df4b605f84b44 and df3bc5a752f0dfcd03393c5f62409814722e720b

File: components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/services/TemplateManagerService.java (hunk 5 above)
Comment (AnuGayan) [reply components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/services/TemplateManagerService.java]: Fixed by 47e9832228beaa5747e074303ac9d729cc69abdf

File: components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/services/TemplateManagerService.java (hunk 6 above)
Comment (AnuGayan) [reply components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/services/TemplateManagerService.java]: Fixed by 47e9832 , 84ec927 and 9a294d6

File: components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/services/TemplateManagerService.java (hunk 7 above)
Comment (AnuGayan) [reply components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/services/TemplateManagerService.java]: Fixed by 47e9832228beaa5747e074303ac9d729cc69abdf

File: components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/services/TemplateManagerService.java (hunk 8 above)
Comment (AnuGayan) [reply components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/services/TemplateManagerService.java]: Fixed by 47e9832228beaa5747e074303ac9d729cc69abdf

File: components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/services/TemplateManagerService.java (hunk 9 above)
Comment (AnuGayan) [reply components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/services/TemplateManagerService.java]: Fixed by 47e9832228beaa5747e074303ac9d729cc69abdf

File: components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/services/TemplateManagerService.java (hunk 10 above)
Comment (AnuGayan) [reply components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/services/TemplateManagerService.java]: Fixed it by ba97bd5e5ac2fa22efaa3d0a89e7024ac8587491

File: components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/util/TemplateManagerHelper.java (hunk 11 above)
Comment (AnuGayan) [reply components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/util/TemplateManagerHelper.java]: FIxed by f5b95d2e8cbf3ac583081822033d065ab10780f5

File: components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/services/TemplateManagerService.java (hunk 4 above)
Comment (AnuGayan) [reply components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/services/TemplateManagerService.java]: Fixed by ba97bd5e5ac2fa22efaa3d0a89e7024ac8587491

File: components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/services/TemplateManagerService.java (hunk 15 above)
Comment (AnuGayan) [reply components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/services/TemplateManagerService.java]: Fixed from df3bc5a752f0dfcd03393c5f62409814722e720b

File: components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/datasource/QueryExecutor.java (hunk 12 above)
Comment (AnuGayan) [reply components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/datasource/QueryExecutor.java]: Fixed by bd9e014c11be3332f023a4e05e13149c5147ebb7

File: components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/datasource/QueryExecutor.java (hunk 13 above)
Comment (AnuGayan) [reply components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/datasource/QueryExecutor.java]: Fixed by bd9e014

**Pull request merged**